### PR TITLE
Oop Homework V3

### PR DESCRIPTION
**Pull request by git-ebetts**
Completed through solution 2. Created tests for Solution 1 (before assignment shifted).

### CODE REVIEW

**Comment by git-ebetts** [comment]
OOP HOMEWORK DUE FOR CLASS 4 and a QUESTION:
***\* May I ask you to review how to set tests up for asynchronous execution? Syntax presented in documentation is not clear. I think that's why these are not passing consistently. ****

**Comment by git-ebetts** [comment]
V3 is all in place for both files but I'm continuing to fiddle with tests -- to make them function asynchronously. See previous note. Thank you!

**Comment by git-ebetts** [comment]
OOP HOMEWORK DUE FOR CLASS 4:
Tests now fully functional. Done!

**Comment by git-ebetts** [comment]
Added child constructor. Closer to a completed V4!

File: js/super_hero.js [hunk 1]
```diff
@@ -1 +1,182 @@
...
+var h = "GOOD TRUMPS EVIL AGAIN!!";
+var j = "NOT SO FAST! THIS BOMB IS THE END TO ALL YOUR EVIL";
+
+
+ 
+ //Method using random number generator to determine 
+ //which properties are active on instance of SuperHero
+ SuperHero.prototype.attack = function (){
+ 
+ 	if (Math.round(Math.random() * 18) % 2 === 0) {
```
Comment (afeld) [inline js/super_hero.js:86]: The logic is sound here, but it wouldn't be immediately obvious to a reader what it's actually checking. How would you describe the condition, in plain English? Then, make a method with that name, and it will be easier for the reader to follow. Something like `if (chance(2, 18)) { ...`.

File: js/super_hero.js [hunk 2]
```diff
@@ -1 +1,182 @@
...
+ 		this.hasSecretPower = false;
+ 	}
+ 
+ 	else if (Math.round(Math.random() * 25) % 5 === 0) {
+ 		this.hasWeapon = false;
+ 	}
+
+ 	if (this.hasBomb === true) {
```
Comment (afeld) [inline js/super_hero.js:94]: Though this is more explicit, you'll more commonly see `if (this.hasBomb)` and `if (!this.hasSecretPower)`, just for conciseness.

File: js/tests.js [hunk 3]
```diff
@@ -1,5 +1,186 @@
...
+
+	} else if
+
+		(m.hasSecretPower === true && m.hasWeapon === true) {
+			assert.strictEqual(attack, c, 'With these secret powers you don\'t have a chance!!');
+
+	} else
+
+		{assert.strictEqual(attack, c, 'With these secret powers you don\'t have a chance!!');
+	}
```
Comment (afeld) [inline js/tests.js:94]: Hmm, yeah, tough to test random behavior. When that's the case, you will often see people [stubbing](https://en.wikipedia.org/wiki/Mock_object) methods that have non-deterministic behavior (using a library like [Sinon](http://sinonjs.org/)), so that their tests _can_ be deterministic.

Comment (afeld) [inline js/tests.js:94]: Does that make sense? Didn't get to cover those kinds of strategies in class, but definitely useful to know about.

**Comment by afeld** [comment]
Great work!

> May I ask you to review how to set tests up for asynchronous execution? ... I think that's why these are not passing consistently.

So sorry for not seeing this sooner – looks like you figured out a way around it? From a quick glance, it doesn't appear that it was an issue of the tests running asynchronously, but instead the random behavior, which I just mentioned above. Let me know if you have any other questions on it!
